File: js/src/events.js
/**
 * Event generators for door openings, fault injection, and alarm derivation.
 *
 * Port of ccesim/simulator/events.py
 */

import { FaultType } from "./config.js";

/**
 * A single door opening event within a sample interval.
 * @typedef {Object} DoorEvent
 * @property {number} start_offset_s - Seconds from interval start when door opens.
 * @property {number} duration_s - Duration of the door opening in seconds.
 */

/**
 * Create a DoorEvent object.
 * @param {number} start_offset_s
 * @param {number} duration_s
 * @returns {DoorEvent}
 */
export function DoorEvent(start_offset_s, duration_s) {
  return { start_offset_s, duration_s };
}

/**
 * The effects of an active fault on the simulation.
 */
export class FaultEffects {
  /**
   * @param {Object} [opts]
   * @param {boolean} [opts.compressor_available=true]
   * @param {boolean} [opts.door_forced_open=false]
   * @param {number} [opts.q_compressor_multiplier=1.0]
   * @param {boolean|null} [opts.power_available_override=null]
   */
  constructor({
    compressor_available = true,
    door_forced_open = false,
    q_compressor_multiplier = 1.0,
    power_available_override = null,
  } = {}) {
    this.compressor_available = compressor_available;
    this.door_forced_open = door_forced_open;
    this.q_compressor_multiplier = q_compressor_multiplier;
    this.power_available_override = power_available_override;
  }
}

/**
 * Generates door opening events using a non-homogeneous Poisson process.
 */
export class DoorEventGenerator {
  /**
   * @param {import('./config.js').EventConfig} config
   * @param {import('./random.js').SeededRandom} rng
   */
  constructor(config, rng) {
    this.config = config;
    this.rng = rng;
  }

  /**
   * Check whether a timestamp falls within working hours.
   * @param {Date} timestamp
   * @returns {boolean}
   */
  _isWorkingHours(timestamp) {
    const hour = timestamp.getUTCHours() + timestamp.getUTCMinutes() / 60.0;
    const [start, end] = this.config.working_hours;
    return start <= hour && hour < end;
  }

  /**
   * Generate door events for one sample interval.
   * @param {Date} intervalStart
   * @param {number} interval_s - Interval length in seconds.
   * @returns {DoorEvent[]}
   */
  getDoorEvents(intervalStart, interval_s) {
    const rate = this._isWorkingHours(intervalStart)
      ? this.config.door_rate_per_hour
      : this.config.door_rate_per_hour * this.config.off_hours_rate_fraction;

    const expected = rate * (interval_s / 3600.0);
    const nEvents = this.rng.poisson(expected);

    const events = [];
    for (let i = 0; i < nEvents; i++) {
      const offset = this.rng.uniform(0, interval_s);
      let duration = Math.max(
        1.0,
        this.rng.gauss(
          this.config.door_mean_duration_s,
          this.config.door_std_duration_s,
        ),
      );
      // Clamp so it doesn't extend beyond interval
      duration = Math.min(duration, interval_s - offset);
      events.push(DoorEvent(offset, duration));
    }

    return events;
  }
}

/**
 * Manages fault state and computes fault effects over time.
 */
export class FaultInjector {
  /**
   * @param {import('./config.js').FaultConfig} config
   * @param {Date} simStart
   */
  constructor(config, simStart) {
    this.config = config;
    this.simStart = simStart;

    if (config.fault_type !== FaultType.NONE) {
      this.faultStart = new Date(
        simStart.getTime() + config.fault_start_offset_s * 1000,
      );
      if (config.fault_duration_s > 0) {
        this.faultEnd = new Date(
          this.faultStart.getTime() + config.fault_duration_s * 1000,
        );
      } else {
        this.faultEnd = null; // Permanent fault
      }
    } else {
      this.faultStart = null;
      this.faultEnd = null;
    }
  }

  /**
   * @param {Date} timestamp
   * @returns {boolean}
   */
  isFaultActive(timestamp) {
    if (this.config.fault_type === FaultType.NONE) return false;
    if (this.faultStart === null) return false;
    if (timestamp.getTime() < this.faultStart.getTime()) return false;
    if (
      this.faultEnd !== null &&
      timestamp.getTime() >= this.faultEnd.getTime()
    )
      return false;
    return true;
  }

  /**
   * Get the effects of any active fault at the given timestamp.
   * @param {Date} timestamp
   * @returns {FaultEffects}
   */
  getFaultEffects(timestamp) {
    if (!this.isFaultActive(timestamp)) {
      return new FaultEffects();
    }

    const fault = this.config.fault_type;

    if (fault === FaultType.POWER_OUTAGE) {
      return new FaultEffects({
        compressor_available: false,
        power_available_override: false,
      });
    } else if (fault === FaultType.STUCK_DOOR) {
      return new FaultEffects({ door_forced_open: true });
    } else if (fault === FaultType.COMPRESSOR_FAILURE) {
      return new FaultEffects({ compressor_available: false });
    } else if (fault === FaultType.REFRIGERANT_LEAK) {
      const elapsed_h =
        (timestamp.getTime() - this.faultStart.getTime()) / 1000 / 3600.0;
      let multiplier = Math.exp(
        -this.config.refrigerant_leak_rate * elapsed_h,
      );
      // Below 5% capacity, the syphon can't sustain two-phase flow
      if (multiplier < 0.05) {
        multiplier = 0.0;
      }
      return new FaultEffects({ q_compressor_multiplier: multiplier });
    }

    return new FaultEffects();
  }
}

/**
 * Derives alarm fields from simulation state.
 *
 * Alarm codes follow the WHO PQS E003 specification.
 *
 * Supported alarms and their continuous-excursion thresholds:
 * - HEAT: TVC > +8 C for 10 continuous hours.
 * - FRZE: TVC <= -0.5 C for 60 continuous minutes.
 * - DOOR: door/lid continuously open for >= 5 minutes.
 * - POWR: continuous no-power condition for >= 24 hours.
 */
export class AlarmGenerator {
  /** Duration thresholds (seconds) */
  static HEAT_THRESHOLD_S = 10 * 3600; // 10 hours
  static FRZE_THRESHOLD_S = 60 * 60; // 60 minutes
  static DOOR_THRESHOLD_S = 5 * 60; // 5 minutes
  static POWR_THRESHOLD_S = 24 * 3600; // 24 hours

  // Pre-defined EMD/Logger error codes (PQS E006 DS01.2 Annex-1 "Error Codes").
  // Codes are 1-7 chars; multiple active codes are space-separated (e.g.
  // "COMM PWR"). EMD errors include the compressor diagnostic codes 1-7 plus
  // the communication code; Logger errors cover sensor/battery/self-test faults.
  static EMD_ERROR_CODES = ["1", "2", "3", "4", "5", "6", "7", "COMM"];
  static LOGGER_ERROR_CODES = ["COMM", "SENS", "BATT", "RTC", "MEM"];

  /**
   * @param {import('./random.js').SeededRandom} rng
   */
  constructor(rng) {
    this.rng = rng;
    /** @type {Date|null} */
    this._lastPowerLoss = null;
    this._powerWasAvailable = true;
    // Continuous excursion tracking
    /** @type {Date|null} */
    this._heatExcursionStart = null;
    /** @type {Date|null} */
    this._frzeExcursionStart = null;
    // Cross-interval door continuity tracking
    /** @type {Date|null} */
    this._continuousDoorStart = null;
    this._doorOpenAtPrevEnd = false;
  }

  /**
   * Merge overlapping/adjacent door events into continuous spans.
   * @param {DoorEvent[]} doorEvents
   * @returns {Array<[number, number]>} Sorted list of [start_offset_s, end_offset_s]
   */
  _mergeDoorSpans(doorEvents) {
    if (!doorEvents || doorEvents.length === 0) return [];
    const EPSILON = 1.0;
    const events = [...doorEvents].sort(
      (a, b) => a.start_offset_s - b.start_offset_s,
    );
    const spans = [];
    let curStart = events[0].start_offset_s;
    let curEnd = events[0].start_offset_s + events[0].duration_s;
    for (let i = 1; i < events.length; i++) {
      const e = events[i];
      const eEnd = e.start_offset_s + e.duration_s;
      if (e.start_offset_s <= curEnd + EPSILON) {
        curEnd = Math.max(curEnd, eEnd);
      } else {
        spans.push([curStart, curEnd]);
        curStart = e.start_offset_s;
        curEnd = eEnd;
      }
    }
    spans.push([curStart, curEnd]);
    return spans;
  }

  /**
   * Evaluate DOOR alarm using sub-interval door events.
   * Detects continuous door-open periods >= DOOR_THRESHOLD_S, including
   * spans that straddle interval boundaries.
   * @param {DoorEvent[]} doorEvents
   * @param {number} interval_s
   * @param {Date} timestamp
   * @returns {boolean}
   */
  _checkDoorAlarm(doorEvents, interval_s, timestamp) {
    const EPSILON = 1.0;
    const spans = this._mergeDoorSpans(doorEvents);

    if (spans.length === 0) {
      this._continuousDoorStart = null;
      this._doorOpenAtPrevEnd = false;
      return false;
    }

    const openAtStart = spans[0][0] <= EPSILON;
    const openAtEnd = spans[spans.length - 1][1] >= interval_s - EPSILON;

    let triggered = false;

    if (this._doorOpenAtPrevEnd && openAtStart) {
      // Continuity from previous interval
      const continuousEnd = spans[0][1];
      const totalS =
        (timestamp.getTime() - this._continuousDoorStart.getTime()) / 1000 +
        continuousEnd;
      if (totalS >= AlarmGenerator.DOOR_THRESHOLD_S) {
        triggered = true;
      }
      // If the first span doesn't reach interval end, the carry-over chain breaks
      if (!openAtEnd || (spans.length > 1 && spans[0][1] < interval_s - EPSILON)) {
        if (openAtEnd) {
          // A *different* span reaches the end -- new chain
          const lastStart = spans[spans.length - 1][0];
          this._continuousDoorStart = new Date(
            timestamp.getTime() + lastStart * 1000,
          );
        } else {
          this._continuousDoorStart = null;
        }
      }
      // else: single span covers start-to-end, keep existing _continuousDoorStart
    } else {
      // No carry-over -- check individual spans within this interval
      for (const [sStart, sEnd] of spans) {
        if (sEnd - sStart >= AlarmGenerator.DOOR_THRESHOLD_S) {
          triggered = true;
          break;
        }
      }
      // Seed carry-over from the span that reaches the interval end
      if (openAtEnd) {
        this._continuousDoorStart = new Date(
          timestamp.getTime() + spans[spans.length - 1][0] * 1000,
        );
      } else {
        this._continuousDoorStart = null;
      }
    }

    this._doorOpenAtPrevEnd = openAtEnd;
    return triggered;
  }

  /**
   * Pick one (usually) or two distinct pre-defined error codes, returned as a
   * space-separated string per the Annex format for multiple codes (e.g.
   * "COMM PWR").
   * @param {string[]} codes
   * @returns {string}
   */
  _pickErrorCodes(codes) {
    const n = Math.min(this.rng.random() < 0.8 ? 1 : 2, codes.length);
    const pool = codes.slice();
    const chosen = [];
    for (let i = 0; i < n; i++) {
      const idx = Math.floor(this.rng.random() * pool.length);
      chosen.push(pool.splice(idx, 1)[0]);
    }
    return chosen.join(" ");
  }

  /**
   * Compute ALRM, EERR, and LERR fields.
   *
   * HOLD (holdover autonomy) is NOT derived here. It is a thermal-reserve
   * estimate reported continuously, computed by the thermal model from the
   * icebank reserve -- not a power-outage stopwatch keyed off
   * power_available.
   *
   * @param {Object} opts
   * @param {number} opts.tvc - Current vaccine chamber temperature.
   * @param {boolean} opts.power_available - Whether power is currently available.
   * @param {Date} opts.timestamp - Current timestamp.
   * @param {DoorEvent[]|null} [opts.door_events=null] - Sub-interval door events.
   * @param {number} [opts.interval_s=900.0] - Length of the sample interval in seconds.
   * @returns {{ALRM: string|null, EERR: string|null, LERR: string|null}}
   */
  deriveAlarms({ tvc, power_available, timestamp, door_events = null, interval_s = 900.0 }) {
    // Track power loss for the POWR continuous-outage alarm.
    if (!power_available && this._powerWasAvailable) {
      this._lastPowerLoss = timestamp;
    }
    if (power_available) {
      this._lastPowerLoss = null;
    }
    this._powerWasAvailable = power_available;

    // Collect active alarm codes
    const codes = [];

    // HEAT: continuous TVC > 8.0 C for 10 hours
    if (tvc > 8.0) {
      if (this._heatExcursionStart === null) {
        this._heatExcursionStart = timestamp;
      }
      const elapsed =
        (timestamp.getTime() - this._heatExcursionStart.getTime()) / 1000;
      if (elapsed >= AlarmGenerator.HEAT_THRESHOLD_S) {
        codes.push("HEAT");
      }
    } else {
      this._heatExcursionStart = null;
    }

    // FRZE: continuous TVC <= -0.5 C for 60 minutes
    if (tvc <= -0.5) {
      if (this._frzeExcursionStart === null) {
        this._frzeExcursionStart = timestamp;
      }
      const elapsed =
        (timestamp.getTime() - this._frzeExcursionStart.getTime()) / 1000;
      if (elapsed >= AlarmGenerator.FRZE_THRESHOLD_S) {
        codes.push("FRZE");
      }
    } else {
      this._frzeExcursionStart = null;
    }

    // DOOR: continuous door open >= 5 minutes
    if (door_events !== null && this._checkDoorAlarm(door_events, interval_s, timestamp)) {
      codes.push("DOOR");
    }

    // POWR: continuous no-power >= 24 hours
    if (this._lastPowerLoss !== null) {
      const powerElapsed =
        (timestamp.getTime() - this._lastPowerLoss.getTime()) / 1000;
      if (powerElapsed >= AlarmGenerator.POWR_THRESHOLD_S) {
        codes.push("POWR");
      }
    }

    // EERR: low-probability EMD error code, drawn from the Annex set.
    let eerr = null;
    if (this.rng.random() < 0.001) {
      eerr = this._pickErrorCodes(AlarmGenerator.EMD_ERROR_CODES);
    }

    // LERR: low-probability logger error code, drawn independently of EERR
    // from the Annex set -- logger and EMD errors are unrelated conditions.
    let lerr = null;
    if (this.rng.random() < 0.001) {
      lerr = this._pickErrorCodes(AlarmGenerator.LOGGER_ERROR_CODES);
    }

    return {
      ALRM: codes.length > 0 ? codes.join(" ") : null,
      EERR: eerr,
      LERR: lerr,
    };
  }
}

File: ccesim/simulator/events.py
"""
Event generators for door openings, fault injection, and alarm derivation.
"""

import math
import random
import datetime as dt
from dataclasses import dataclass
from typing import List, Optional

from ccesim.simulator.config import EventConfig, FaultConfig, FaultType
from ccesim.simulator.thermal import DoorEvent


@dataclass
class FaultEffects:
    """The effects of an active fault on the simulation.

    compressor_available: Whether the compressor can run.
    door_forced_open: Whether the door is stuck open.
    q_compressor_multiplier: Multiplier on Q_compressor (< 1.0 for refrigerant leak).
    power_available_override: If set, overrides the power model's availability.
    """
    compressor_available: bool = True
    door_forced_open: bool = False
    q_compressor_multiplier: float = 1.0
    power_available_override: Optional[bool] = None


class DoorEventGenerator:
    """Generates door opening events using a non-homogeneous Poisson process."""

    def __init__(self, config: EventConfig, rng: random.Random):
        self.config = config
        self.rng = rng

    def _is_working_hours(self, timestamp: dt.datetime) -> bool:
        hour = timestamp.hour + timestamp.minute / 60.0
        start, end = self.config.working_hours
        return start <= hour < end

    def get_door_events(self, interval_start: dt.datetime, interval_s: float) -> List[DoorEvent]:
        """Generate door events for one sample interval.

        Uses a Poisson process with rate depending on whether we're in working hours.
        """
        if self._is_working_hours(interval_start):
            rate = self.config.door_rate_per_hour
        else:
            rate = self.config.door_rate_per_hour * self.config.off_hours_rate_fraction

        # Expected events in this interval
        expected = rate * (interval_s / 3600.0)
        n_events = self.rng.poisson(expected) if hasattr(self.rng, 'poisson') else self._poisson(expected)

        events = []
        for _ in range(n_events):
            offset = self.rng.uniform(0, interval_s)
            duration = max(1.0, self.rng.gauss(
                self.config.door_mean_duration_s,
                self.config.door_std_duration_s
            ))
            # Clamp duration so it doesn't extend beyond interval
            duration = min(duration, interval_s - offset)
            events.append(DoorEvent(start_offset_s=offset, duration_s=duration))

        return events

    def _poisson(self, lam: float) -> int:
        """Generate a Poisson-distributed random number using the inverse transform."""
        if lam <= 0:
            return 0
        import math
        L = math.exp(-lam)
        k = 0
        p = 1.0
        while True:
            k += 1
            p *= self.rng.random()
            if p < L:
                return k - 1


class FaultInjector:
    """Manages fault state and computes fault effects over time."""

    def __init__(self, config: FaultConfig, sim_start: dt.datetime):
        self.config = config
        self.sim_start = sim_start

        if config.fault_type != FaultType.NONE:
            self.fault_start = sim_start + dt.timedelta(seconds=config.fault_start_offset_s)
            if config.fault_duration_s > 0:
                self.fault_end = self.fault_start + dt.timedelta(seconds=config.fault_duration_s)
            else:
                self.fault_end = None  # Permanent fault
        else:
            self.fault_start = None
            self.fault_end = None

    def is_fault_active(self, timestamp: dt.datetime) -> bool:
        if self.config.fault_type == FaultType.NONE:
            return False
        if self.fault_start is None:
            return False
        if timestamp < self.fault_start:
            return False
        if self.fault_end is not None and timestamp >= self.fault_end:
            return False
        return True

    def get_fault_effects(self, timestamp: dt.datetime) -> FaultEffects:
        """Get the effects of any active fault at the given timestamp."""
        if not self.is_fault_active(timestamp):
            return FaultEffects()

        fault = self.config.fault_type

        if fault == FaultType.POWER_OUTAGE:
            return FaultEffects(
                compressor_available=False,
                power_available_override=False,
            )
        elif fault == FaultType.STUCK_DOOR:
            return FaultEffects(door_forced_open=True)
        elif fault == FaultType.COMPRESSOR_FAILURE:
            return FaultEffects(compressor_available=False)
        elif fault == FaultType.REFRIGERANT_LEAK:
            # Exponential decay of cooling capacity.
            #
            # Real-world thermosyphon leaks (from InfluxDB field data) show:
            #   - Gradual loss over weeks/months, not hours
            #   - Two-phase pattern: degraded-but-stable, then rapid failure
            #     once icebank depletes (the icebank dynamics handle phase 2)
            #   - Exponential fits physics: leak rate ∝ remaining refrigerant
            #     pressure, which decreases as gas escapes
            #
            # Reference case from the deployed fleet: onset early March, a
            # 60-day degradation, then failure once the icebank depleted.
            elapsed_h = (timestamp - self.fault_start).total_seconds() / 3600.0
            multiplier = math.exp(-self.config.refrigerant_leak_rate * elapsed_h)
            # Below 5% capacity, the syphon can't sustain two-phase flow
            # and effectively disconnects — no cooling reaches the VCC.
            if multiplier < 0.05:
                multiplier = 0.0
            return FaultEffects(q_compressor_multiplier=multiplier)

        return FaultEffects()


class AlarmGenerator:
    """Derives alarm fields from simulation state.

    Alarm codes follow the WHO PQS E003 specification.  The ALRM field
    carries all currently active codes as a space-separated string
    (e.g. ``"HEAT DOOR"``).

    Supported alarms and their continuous-excursion thresholds:
    - HEAT: TVC > +8 °C for 10 continuous hours.
    - FRZE: TVC <= −0.5 °C for 60 continuous minutes.
    - DOOR: door/lid continuously open for > 5 minutes.
    - POWR: continuous no-power condition for > 24 hours.
    """

    # Duration thresholds (seconds)
    HEAT_THRESHOLD_S = 10 * 3600   # 10 hours
    FRZE_THRESHOLD_S = 60 * 60     # 60 minutes
    DOOR_THRESHOLD_S = 5 * 60      # 5 minutes
    POWR_THRESHOLD_S = 24 * 3600   # 24 hours

    # Pre-defined EMD/Logger error codes (PQS E006 DS01.2 Annex-1 "Error Codes").
    # Codes are 1-7 chars; multiple active codes are space-separated (e.g.
    # "COMM PWR").  EMD errors include the compressor diagnostic codes 1-7 plus
    # the communication code; Logger errors cover sensor/battery/self-test faults.
    EMD_ERROR_CODES = ("1", "2", "3", "4", "5", "6", "7", "COMM")
    LOGGER_ERROR_CODES = ("COMM", "SENS", "BATT", "RTC", "MEM")

    def __init__(self, rng: random.Random):
        self.rng = rng
        self._last_power_loss: Optional[dt.datetime] = None
        self._power_was_available: bool = True
        # Continuous excursion tracking
        self._heat_excursion_start: Optional[dt.datetime] = None
        self._frze_excursion_start: Optional[dt.datetime] = None
        # Cross-interval door continuity tracking
        self._continuous_door_start: Optional[dt.datetime] = None
        self._door_open_at_prev_end: bool = False

    def _merge_door_spans(self, door_events: List[DoorEvent]) -> List[tuple]:
        """Merge overlapping/adjacent door events into continuous spans.

        Returns a sorted list of (start_offset_s, end_offset_s) tuples.
        """
        if not door_events:
            return []
        EPSILON = 1.0
        events = sorted(door_events, key=lambda e: e.start_offset_s)
        spans = []
        cur_start = events[0].start_offset_s
        cur_end = events[0].start_offset_s + events[0].duration_s
        for e in events[1:]:
            e_end = e.start_offset_s + e.duration_s
            if e.start_offset_s <= cur_end + EPSILON:
                cur_end = max(cur_end, e_end)
            else:
                spans.append((cur_start, cur_end))
                cur_start = e.start_offset_s
                cur_end = e_end
        spans.append((cur_start, cur_end))
        return spans

    def _check_door_alarm(
        self,
        door_events: List[DoorEvent],
        interval_s: float,
        timestamp: dt.datetime,
    ) -> bool:
        """Evaluate DOOR alarm using sub-interval door events.

        Detects continuous door-open periods >= DOOR_THRESHOLD_S, including
        spans that straddle interval boundaries.
        """
        EPSILON = 1.0
        spans = self._merge_door_spans(door_events)

        if not spans:
            # No door activity — reset continuity
            self._continuous_door_start = None
            self._door_open_at_prev_end = False
            return False

        open_at_start = spans[0][0] <= EPSILON
        open_at_end = spans[-1][1] >= interval_s - EPSILON

        triggered = False

        if self._door_open_at_prev_end and open_at_start:
            # Continuity from previous interval — find how far the
            # door stays open from offset 0 (first merged span).
            continuous_end = spans[0][1]
            total_s = (timestamp - self._continuous_door_start).total_seconds() + continuous_end
            if total_s >= self.DOOR_THRESHOLD_S:
                triggered = True
            # If the first span doesn't reach interval end, the carry-over
            # chain breaks here; start fresh from the last span if needed.
            if not open_at_end or (len(spans) > 1 and spans[0][1] < interval_s - EPSILON):
                # First span didn't bridge all the way to end; reset.
                if open_at_end:
                    # A *different* span reaches the end — new chain.
                    last_start = spans[-1][0]
                    self._continuous_door_start = timestamp + dt.timedelta(seconds=last_start)
                else:
                    self._continuous_door_start = None
            # else: single span covers start-to-end, keep existing _continuous_door_start
        else:
            # No carry-over — check individual spans within this interval.
            for s_start, s_end in spans:
                if s_end - s_start >= self.DOOR_THRESHOLD_S:
                    triggered = True
                    break
            # Seed carry-over from the span that reaches the interval end.
            if open_at_end:
                self._continuous_door_start = timestamp + dt.timedelta(seconds=spans[-1][0])
            else:
                self._continuous_door_start = None

        self._door_open_at_prev_end = open_at_end
        return triggered

    def _pick_error_codes(self, codes) -> str:
        """Pick one (usually) or two distinct pre-defined error codes.

        Returns a space-separated string, per the Annex "Data Objects" format
        for multiple simultaneous codes (e.g. "COMM PWR").
        """
        n = min(1 if self.rng.random() < 0.8 else 2, len(codes))
        chosen = self.rng.sample(list(codes), n)
        return ' '.join(chosen)

    def derive_alarms(
        self,
        tvc: float,
        power_available: bool,
        timestamp: dt.datetime,
        door_events: Optional[List[DoorEvent]] = None,
        interval_s: float = 900.0,
    ) -> dict:
        """Compute ALRM, EERR, and LERR fields.

        Args:
            tvc: Current vaccine chamber temperature.
            power_available: Whether power is currently available.
            timestamp: Current timestamp.
            door_events: Sub-interval door events for DOOR alarm evaluation.
            interval_s: Length of the sample interval in seconds.

        Returns:
            Dict with ALRM, EERR, LERR keys.  ALRM is a space-separated
            string of all active alarm codes, or None if no alarms are active.
            EERR (EMD error codes) and LERR (logger error codes) are distinct
            manufacturer-defined fields and are drawn independently.

        Note: HOLD (holdover autonomy) is NOT derived here.  It is a
        thermal-reserve estimate reported continuously, computed by the
        thermal model from the icebank reserve -- not a power-outage
        stopwatch keyed off ``power_available``.
        """
        # Track power loss for the POWR continuous-outage alarm.
        if not power_available and self._power_was_available:
            self._last_power_loss = timestamp
        if power_available:
            self._last_power_loss = None
        self._power_was_available = power_available

        # Collect active alarm codes
        codes: List[str] = []

        # HEAT: continuous TVC > 8.0°C for 10 hours
        if tvc > 8.0:
            if self._heat_excursion_start is None:
                self._heat_excursion_start = timestamp
            elapsed = (timestamp - self._heat_excursion_start).total_seconds()
            if elapsed >= self.HEAT_THRESHOLD_S:
                codes.append("HEAT")
        else:
            self._heat_excursion_start = None

        # FRZE: continuous TVC <= -0.5°C for 60 minutes
        if tvc <= -0.5:
            if self._frze_excursion_start is None:
                self._frze_excursion_start = timestamp
            elapsed = (timestamp - self._frze_excursion_start).total_seconds()
            if elapsed >= self.FRZE_THRESHOLD_S:
                codes.append("FRZE")
        else:
            self._frze_excursion_start = None

        # DOOR: continuous door open > 5 minutes
        if door_events is not None and self._check_door_alarm(door_events, interval_s, timestamp):
            codes.append("DOOR")

        # POWR: continuous no-power > 24 hours
        if self._last_power_loss is not None:
            power_elapsed = (timestamp - self._last_power_loss).total_seconds()
            if power_elapsed >= self.POWR_THRESHOLD_S:
                codes.append("POWR")

        # EERR: low-probability EMD error code, drawn from the Annex set.
        eerr = None
        if self.rng.random() < 0.001:
            eerr = self._pick_error_codes(self.EMD_ERROR_CODES)

        # LERR: low-probability logger error code, drawn independently of EERR
        # from the Annex set -- logger and EMD errors are unrelated conditions.
        lerr = None
        if self.rng.random() < 0.001:
            lerr = self._pick_error_codes(self.LOGGER_ERROR_CODES)

        return {
            'ALRM': ' '.join(codes) if codes else None,
            'EERR': eerr,
            'LERR': lerr,
        }
